srml/aura/src/lib.rs

```diff
@@ -154,6 +163,72 @@ pub trait Trait: timestamp::Trait {
 
 	/// The identifier type for an authority.
 	type AuthorityId: Member + Parameter + Default;
+
+	/// The signature type for an authority.
+	type Signature: Verify<Signer = Self::AuthorityId> + Parameter;
+}
+
+fn handle_equivocation_proof<T: Trait>(
+	proof: &AuraEquivocationProof<T::Header, T::Signature>
+) -> TransactionValidity 
+{
+	let header1 = proof.first_header();
+	let header2 = proof.second_header();
+
+	if header1 == header2 {
+		return TransactionValidity::Invalid(0)
+	}
+
+	let maybe_slot1 = find_pre_digest::<_, T::Signature>(header1);
+	let maybe_slot2 = find_pre_digest::<_, T::Signature>(header2);
+
+	if maybe_slot1.is_ok() && maybe_slot1 == maybe_slot2 {
+		let slot = maybe_slot1.expect("OK by previous line; qed");
+
+		// FIX: What if there is a different set of authorities?
+		let authorities = <Module<T>>::authorities();
```

Comment:
right, this is the hardest part of slashing. we probably will need the historical session trie to also store some key mapped to `Vec<T::Keys>`. Consensus engines that need access to the entire authorities list will have to accept a proof of that branch as well.